### PR TITLE
Clarify criteria for maintainer nominations

### PR DESCRIPTION
**Pull request by denyeart**
Clarify criteria for maintainer nominations

Signed-off-by: David Enyeart <enyeart@us.ibm.com>

### CODE REVIEW

File: docs/source/CONTRIBUTING.rst
Comment (mastersingh24) [inline docs/source/CONTRIBUTING.rst]: Do we want to better define inactivity or just leave that up to the judgement of the maintainers at the time a proposal for removal?

**Review by mastersingh24 — DISMISSED** [review]

File: docs/source/CONTRIBUTING.rst
Comment (denyeart) [reply docs/source/CONTRIBUTING.rst]: I've clarified as "e.g. 3 or more months with no review comments". This way there is at least one measurable criteria based on maintainer's primary responsibility, while still leaving the door open for judgement.

**Review by yacovm — APPROVED** [review]
timer.Reset(time.Hour * 24 * 31 * 3)

**Review by christo4ferris — APPROVED** [review]
Good change. I might have characterized reviews as being 'constructive quality reviews', but this works

File: docs/source/CONTRIBUTING.rst
Comment (MHBauer) [inline docs/source/CONTRIBUTING.rst:125]: Is the majority required? Is it those who show up to whatever vote, or majority of all in the list?

**Review by MHBauer — CHANGES_REQUESTED** [review]
I'm familiar with participating in some other projects, for example:

[Docker](https://github.com/moby/moby/blob/8e610b2b55bfd1bfa9436ab110d311f5e8a74dcb/project/GOVERNANCE.md)
[Kubernetes](https://github.com/kubernetes/community/blob/master/community-membership.md)
[more kubernetes](https://github.com/kubernetes/community/blob/master/contributors/guide/owners.md)

I don't think we have the metrics to take a "count the contributions" approach as kubernetes.

The current proposal does not clarify anything over the current text, and seems to be a very high bar.

My preference is to be as inclusive as possible as fast as possible, but I understand this project requires more stability than some others.
Docker has something closer to the current proposal, in terms of "you get it when you earn the trust of others". 

Concrete suggestions:
 - More measurable criteria. Including a dashboard or some such to check your progress and try to gameify behavior.
 - Acknowledgement that "maintainers feelings on your behavior" come into decisions.
 - More use of the codowners file to segregate responsibilities.
 - Addition of some minor additional hierarchy, perhaps a veto power.

File: docs/source/CONTRIBUTING.rst
Comment (cendhu) [reply docs/source/CONTRIBUTING.rst:112]: @MHBauer I think you expect something very concrete like what is mentioned in TiDB community and I do agree with you. For e.g., a PR can be reviewed using golang coding guidelines without actually having the deep technical knowledge or familiarity with the codebase (which is necessary IMO). Some of the concrete items mentioned in the below list can be used here. Btw, it is not necessary to tick all the mentioned points as we cannot expect everyone to be perfect on every task. Sometimes, the opportunity itself would not be there to showcase their skills in a particular item. 


**Contributor** 
- Community contributors with at least one PR merged to any repository under PingCAP. 

**Active Contributor**

*How to become a reviewer?*
- Contributed 8 PRs and above to TiDB within one year
- Received recognition from two incumbent Reviewers

**Reviewer**

Reviewers are responsible for reviewing the code submitted by the community to ensure the quality and correctness; generated from Contributors.

*How to become a reviewer?*
- Familiar with the codebase
- Contributed a certain amount of code to TiDB subprojects
- Performed excellently in the Special Interest Group (SIG)
- Sponsored by at least two incumbent Committers

*Privileges and responsibilities*
- Ensure the correctness of the subproject code
- Respond promptly to PRs submitted by the community
- Participate in SIG seminars on a regular basis
- Approve code reviews with LGTM 

**Committer**

Committers are recommended by Maintainers or the PMC as Contributors with outstanding contributions to TiDB. A Committer must have completed at least one feature or fix a major bug independently.

*How to become a committer?*
- Demonstrated good technical judgment
- Assumed the Reviewer role in at least two or more subprojects
- Reviewed or merged a certain amount of TiDB code
- Contributed significantly to TiDB performance improvements, or fix major bugs

*Privileges and responsibilities*
- Ensure overall code quality of the project
- Guide contributors and reviewers to contribute to TiDB continuously
- Participate in Committer meetings on a regular basis

**Maintainer**

Maintainers are the planners and designers of the project, with the authority to merge branches into the master. Maintainers come from Committers. They must demonstrate both good judgment and responsibility towards the health of subproject. Maintainers must set up the technical direction and make or approve design decisions for the subproject - either directly or by delegating these responsibilities.

*How to become a maintainer?*
- Demonstrated a deep understanding of TiDB's technical goals and directions
- Actively involved in organizing technical seminars and propose constructive subprojects
- Can handle complex problems in the code implementation process

*Privileges and responsibilities*
- Develop technical design and decisions for subprojects
- Set the technical directions and priorities for subprojects
- Participate in product releases and roadmap development of the project
- Continuously contribute to the project by mentoring Contributors, Reviewers, and Committers 
- Ensure overall code quality of the subprojects

File: docs/source/CONTRIBUTING.rst
Comment (mastersingh24) [reply docs/source/CONTRIBUTING.rst:125]: > Is the majority required? Is it those who show up to whatever vote, or majority of all in the list?

Traditionally, it's been a majority of the listed maintainers.  We generally do not have formal calls / meetings and have handled via Gerrit CRs and now Github PRs.

But it's a good point ... we probably should define something like a required quorum.

**Comment by denyeart** [comment]
@cendhu @MHBauer you have good points, but they go beyond the scope of the intended PR. The PR intentionally leaves it as a judgement call, and simply highlights a few behaviors that maintainers currently consider. There is no policy change intended here, but let's wait a few more days to see if any maintainers object to anything before merging. 

I expect the types of policy changes you are suggesting will result in much more debate and require a vote among maintainers, you are of course welcome to submit a PR for consideration.

**Comment by cendhu** [comment]
@denyeart I agree that it is highly debatable. 

> The PR intentionally leaves it as a judgement call, and simply highlights a few behaviors that maintainers currently consider.

seems to contradict with 

> let's wait a few more days to see if any maintainers object to anything before merging.

:-)

**Review by jyellick — APPROVED** [review]

**Review by mastersingh24 — APPROVED** [review]

**Comment by denyeart** [comment]
Discussion has settled with no objections, I'll go ahead and merge...

**Pull request merged**
